### PR TITLE
Reworked chapter on what is a state machine

### PR DESCRIPTION
**Pull request by tkutz**

### CODE REVIEW

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 1]
```diff
@@ -3,81 +3,86 @@ p.
 
 h1(#overview_what_are_state_machines). What is a state machine?
 
-A state machine is a behavior model. It consists of different so called states. Because the amount of states of a state machine is finite it is called finite state machine (FSM). There are basic types like Mealy and Moore machines and more complex types like Harel and UML statecharts. 
+A state machine is a behavior model. It consists of different so called states. Because the amount of states of a state machine is finite it is called finite state machine (FSM). There are basic types like Mealy and Moore machines and more complex types like Harel and UML statecharts.
 
 This introduction gives a short overview of the common basis and the differences between state machine types.
 
-!(large)images/simple_moore.png(Simple example as a Moore machine)!
+!(small)images/simple_moore.png(Simple example as a Moore machine)!
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: :+1:

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 2]
```diff
@@ -3,81 +3,86 @@ p.
...
 
 p=. Simple state machine
 
-In automata theory a state is a situation of a system depending on previous inputs and causes a reaction on following inputs. Automata theory uses the terms input/output for simple text input/output. Modern state machines use an extended definition of inputs and outputs. These can be events or more complex reactions than simply printing a certain text. Another important element of  state machines are transitions. A transition, as the name says, describes the transition from one state to another.
+The basic building blocks of a state machine are states and transitions. A state is a situation of a system depending on previous inputs and causes a reaction on following inputs. Transitions define for which external input and under which condition a state change occurs.
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: I stay at my point: I would not change the old introduction.

This is simple and on point: "A transition, as the name says, describes the transition from one state to another."
For example, this is how Wikipedia describes it: "the change from one state to another is called a **transition"

This leaves questions: "transitions define for which external input and under which condition a state change occurs."
- What's the difference between external and internal? Why external?
- What is a state change?

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 3]
```diff
@@ -3,81 +3,86 @@ p.
...
 
-To explain this concept we introduce a common example:
+Consider the simple state machine above. It consists of two states, *Off* and *On*. *Off* is the initial state, i.e. it is activated when the state machine is executed. The arrows between the states denote the possible state transitions. They define for which input a state change occurs. Here, the active state is changed from *Off* to *On* for the input *buttonpressed*, and back again to *Off* for the same input.
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: "*Off* is the initial state" -> This is wrong!
The initial state is the "On" state. We have to rework the picture!

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 4]
```diff
@@ -3,81 +3,86 @@ p.
...
 
-The state diagram has four states: *NorthernPort*, *SouthernPort*, *Northwards* and *Southwards*. Each of these states produces an output which is an action of the ferry. The ferry can either stop when one of the two ports is reached (_ferry.stop_) or move from one port to another (_ferry.moveSouthwards_ and _ferry.moveNorthwards_). In Moore machines the output *only* relies on the current state. This means that the output will not change unless the state also changes. This happens when a transition is taken. In our example, we have two kinds of inputs that may trigger a state transition: The ship’s mate who can decide where to sail (e.g. _mate.sailSouthwards_) and the quay which can detect an arriving ferry (e.g. _quay.reachedSouthernBank_).
+We extend the switch example above into a light switch with different brightness levels. The light switch has two buttons, an _ON_ button and an _OFF_ button. Pressing the _ON_ button switches the light on and toggles through the different brightness levels. The output of the state machine is the brightness value. As in Moore machines only states produce outputs, we need one dedicated state per brightness level:
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: "Pressing the _ON_ button switches the light on and toggles through the different brightness levels"

This is not 100% correct. The _ON_ button raises the *ON_pressed* event. It does not turn on or off the light - then we would need no state machine, cause it is simply wired

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 5]
```diff
@@ -3,81 +3,86 @@ p.
...
+p=. Light switch example as a Moore machine, modeled with YAKINDU Statechart Tools
 
-!(large)images/ferry_mealy_1.png(Ferry example as a Mealy machine)!
 
-p=. Ferry example as a Mealy machine
+*Mealy machines*
 
-States do not produce outputs anymore, but the transitions do. Whether the ferry moves northwards or southwards is now expressed by the corresponding transitions. The input event mate.sail may lead to the output _ferry.moveSouthwards_ or _ferry.moveNorthwards_ depending on the active state. The state machine offers an extended logic: the states *Northwards* and *Southwards* have two transitions that lead to different states depending on the input event: If the event _mate.sail_ is raised in *Northwards* the output is _ferry.moveSouthwards_, if _quay.reachedNorthernBank_ is raised the ferry stops. The same logic as a Moore machine would lead to additional states like the following image shows:
+In comparison with Moore machines, "_Mealy_ machines":https://en.wikipedia.org/wiki/Mealy_machine *produce outputs only on transitions* and not in states. This often results in state diagrams with fewer states because more logic can be put on transitions.
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: "In comparison with Moore machines" -> moore kursiv

Comment (tkutz) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: No, italics font is only used when a term is introduced the first time. Afterwards it's normal font style again.

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 6]
```diff
@@ -3,89 +3,98 @@ p.
...
+We extend the switch example above into a light switch with different brightness levels. The light switch has two buttons, an _ON_ button and an _OFF_ button. Pressing the _ON_ button switches the light on and toggles through the different brightness levels. This behavior is modeled by the state machine below. Pressing a button raises a corresponding event (_ON_pressed_ or _OFF_pressed_) upon which the machine reacts with a state change and corresponding outputs/actions. The output of the state machine is simply the brightness value. As in Moore machines only states produce outputs, we need one dedicated state per brightness level:
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: This describes no Moore FSM. FSM have a fixed output alphabet and do not know the concept of variables.

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 7]
```diff
@@ -3,89 +3,98 @@ p.
...
 
-h2(#overview_mealy). Compact state machines by George H. Mealy
+!(small)images/overview_moore.png(Light switch example as a Moore machine, modeled with YAKINDU Statechart Tools)!
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: update model

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 8]
```diff
@@ -3,89 +3,98 @@ p.
...
+In comparison with Moore machines, "_Mealy_ machines":https://en.wikipedia.org/wiki/Mealy_machine *produce outputs only on transitions* and not in states. This often results in state diagrams with fewer states because more logic can be put on transitions. 
 
-!(large)images/ferry_moore2.png(The ferry as an extended Moore machine)!
+In the light switch example we can now merge all _Light On_ states into one and produce the output on a transition. Here we just increment the brightness value each time the transition is taken. We use the modulo expression to ensure the brightness value stays between 1 and 3.
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: This describes no Mealy FSM. FSM have a fixed output alphabet and do not know the concept of variables.

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 9]
```diff
@@ -3,89 +3,98 @@ p.
...
 
-p=. The ferry as an extended Moore machine
+!(small)images/overview_mealy.png(Light switch example as a Mealy machine)!
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: fix state machine to be a Mealy automaton.

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 10]
```diff
@@ -3,89 +3,98 @@ p.
...
-UML state machines are based on the statechart notation introduced by David Harel. Furthermore, the UML extends the notation of Harel statecharts by object-oriented principles. Mapping this to our ferry example, in UML we can model the possible actions of the ferry as an interface with operations stop, moveNorthwards, moveSouthwards and so on.
+Further concepts like orthogonality, entry/exit actions or history states are left out here for brevity. You can read about them in our "quick reference":../user-guide/quick_ref.html.
+
+*The present age: UML state machines*
+
+UML state machines are based on the statechart notation introduced by David Harel. Furthermore, the UML extends the notation of Harel statecharts by object-oriented principles. Mapping this to our light switch example, in UML we can model the possible actions of the light switch as a type with operations @turnOn()@, @turnOff()@, @setBrightness(value)@ and so on.
 
 The following table illustrates the differences between the previously described types at a glance:
 
 !(large)images/statemachine_types.png(Differences between state machine types)!
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: Graphics must list concept of (extended state) variables - checked for Harel and UML

Comment (tkutz) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: Isn't it subsumed in the last line of the table in the term "actions"?

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 11]
```diff
@@ -24,7 +24,7 @@ In the following, we will extend the simple switch example to explain the differ
 
 In automata theory, there are two basic types of finite-state machines (FSM). One of those is called "_Moore_ machine":https://en.wikipedia.org/wiki/Moore_machine, named after its inventor Edward Moore. Moore machines consist of states and transitions. *States are able to produce outputs*, and the output is determined solely by the current state, not by any input.
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: add ?: invented 1956

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 12]
```diff
@@ -3,89 +3,95 @@ p.
...
+
+Using composite states and sub diagrams, we are able to bring more *depth* into state diagrams, while keeping the diagrams clear and well-structured. Regions help us to express *orthogonality*: Different substate machines that can be executed side by side. Events allow us to achieve *broadcast communication* and give us a strong means to describe complex behavior. Using guards, we can state that a certain event triggers a transition only if a given condition is met. Inter-level transitions, history states, temporal logic as well as entry, exit and throughout actions are further Harel statechart elements.
 
-h3(#overview_harel_vs_mealy). Harel vs Mealy
+Since Harel statecharts allow to use variable assignments as output actions, we can simplify the light switch example by merging all _Light On_ states into one and executing the output actions on a self-transition. Here we just increment the brightness value each time the transition is taken. We use the modulo expression to ensure the brightness value stays between 1 and 3.
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: Sketch the problem of state explosion ...

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 13]
```diff
@@ -3,89 +3,95 @@ p.
...
+UML state machines are based on the statechart notation introduced by David Harel. Furthermore, the UML extends the notation of Harel statecharts by object-oriented principles. Mapping this to our light switch example, in UML we can model the possible actions of the light switch as a type with operations @turnOn()@, @turnOff()@, @setBrightness(value)@ and so on.
 
 The following table illustrates the differences between the previously described types at a glance:
 
 !(large)images/statemachine_types.png(Differences between state machine types)!
 
 p=. Differences between the state machine types
 
-The examples of this article were designed with the YAKINDU Statechart Tools, whose documentation you are reading just now.
+The examples of this article were designed with "YAKINDU Statechart Tools":https://www.itemis.com/en/yakindu/state-machine/, whose documentation you are reading just now.
```
Comment (terfloth) [inline plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: Mention to what category YSCT belongs. YAKINDU statecharts are based on Harel stetcharts and are very close but not identical to UML state machines. The concrete difference are explained in the documentation where they exist.

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile [hunk 14]
```diff
@@ -3,89 +3,95 @@ p.
...
 
-!(large)images/simple_moore.png(Simple example as a Moore machine)!
+!(small)images/simple_moore.png(Simple example as a Moore machine)!
 
 p=. Simple state machine
 
-In automata theory a state is a situation of a system depending on previous inputs and causes a reaction on following inputs. Automata theory uses the terms input/output for simple text input/output. Modern state machines use an extended definition of inputs and outputs. These can be events or more complex reactions than simply printing a certain text. Another important element of  state machines are transitions. A transition, as the name says, describes the transition from one state to another.
+The basic building blocks of a state machine are states and transitions. A state is a situation of a system depending on previous inputs and causes a reaction on following inputs. One state is marked as the initial state; this is where the execution of the machine starts. A state transition defines when a state is changed from one to another. This can be an input, like pressing a button, and/or a condition that needs to be fulfilled. Depending on the state machine type, states and/or transitions produce outputs.
```
[the remark this answers is not in the document]
Comment (tkutz) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: ... based on an input.

File: plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile (hunk 13 above)
Comment (tkutz) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/overview.textile]: Simulation, executable state machines...

**Pull request merged**